### PR TITLE
[TorchToTcp] Don't broadcast axes that don't change size

### PR DESCRIPTION
**Pull request by AaronStGeorge**
Ran into this a bug in Cruse internal repo, the changes here fix the issue. Here's an example of incorrect lowering:
```mlir
  %4 = torch.aten.broadcast_to %3, %1 : !torch.vtensor<[1,2,1,2],f32>, !torch.list<int> -> !torch.vtensor<[4,2,4,2],f32>
```
was lowering to 
```mlir
%13 = tcp.broadcast %4, %6, %8, %10, %12 {axes = [0, 1, 2, 3]} : tensor<1x2x1x2xf32>, index, index, index, index -> tensor<4x2x4x2xf32>
```
axes 1 and 3 should not be broadcasted as those axes stay the same shape. The correct lowering is
```mlir
%13 = tcp.broadcast %4, %6, %10 {axes = [0, 2]} : tensor<1x2x1x2xf32>, index, index -> tensor<4x2x4x2xf32>
```

### CODE REVIEW

File: lib/Conversion/TorchToTcp/Misc.cpp [hunk 1]
```diff
@@ -84,7 +86,8 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
       int64_t staticDimSize;
       if (i < newLeadingDims ||
           !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize)) ||
-          staticDimSize != -1) {
+          (staticDimSize != -1 &&
+           staticDimSize != inputShape[i - newLeadingDims])) {
```
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Misc.cpp:90]: perhaps add a short comment above explaining the rationale

File: test/Conversion/TorchToTcp/misc.mlir [hunk 2]
```diff
@@ -322,3 +322,22 @@ func.func @torch.aten.size.int(%arg0: !torch.vtensor<[?,?],f32>) -> () {
...
+// CHECK-LABEL:  @torch.aten.broadcast_to(
+// CHECK-SAME:   %[[ARG:.*]]: !torch.vtensor<[1,2,1,2],f32>) {
+// CHECK:        %[[TENSOR:.*]] = torch_c.to_builtin_tensor %[[ARG]] : !torch.vtensor<[1,2,1,2],f32> -> tensor<1x2x1x2xf32>
+// CHECK:        %[[CONSTANT:.*]] = torch.constant.int 4
+// CHECK:        %[[CAST0:.*]] = torch_c.to_i64 %[[CONSTANT]]
+// CHECK:        %[[BROADCAST_DIM0:.*]] = arith.index_cast %[[CAST0]] : i64 to index
+// CHECK:        %[[CAST1:.*]] = torch_c.to_i64 %[[CONSTANT]]
+// CHECK:        %[[BROADCAST_DIM1:.*]] = arith.index_cast %[[CAST1]] : i64 to index
+// CHECK:        %{{.*}} = tcp.broadcast %[[TENSOR]], %[[BROADCAST_DIM0]], %[[BROADCAST_DIM1]] {axes = [0, 2]} : tensor<1x2x1x2xf32>, index, index -> tensor<4x2x4x2xf32>
+func.func @torch.aten.broadcast_to(%arg0: !torch.vtensor<[1,2,1,2],f32>) -> () {
```
Comment (sjain-stanford) [inline test/Conversion/TorchToTcp/misc.mlir:337]: Thanks for adding the lit test (this was missed earlier). Can you please also onboard this to the `aot_compile` macro for e2e testing? I usually just have chat-GPT write me the pytorch model to produce this Torch dialect program, and feed that into the model loader.

It gave me this:
```python
class MyModel(torch.nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()

    def forward(self, arg0):
        # Define the desired shape as a list
        shape = [4, 2, 4, 2]
        # Broadcast the input tensor to the desired shape
        output = arg0.expand(*shape)
        return output

# Example usage
model = MyModel()
sample_input = torch.randn(1, 2, 1, 2)  # Example input shape [1, 2, 1, 2]
output = model(sample_input)
print(output)
```

File: lib/Conversion/TorchToTcp/Misc.cpp (hunk 1 above)
Comment (navahgar) [reply lib/Conversion/TorchToTcp/Misc.cpp:90]: +1 to this.

How about adding some lambda function that describes what it is actually checking, say something like `isDimBroadcasted` for the new condition you added?

File: lib/Conversion/TorchToTcp/Misc.cpp [hunk 3]
```diff
@@ -84,7 +86,8 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
       int64_t staticDimSize;
       if (i < newLeadingDims ||
           !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize)) ||
-          staticDimSize != -1) {
+          (staticDimSize != -1 &&
```
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Misc.cpp:89]: nit: can we use `ShapedType::kDynamic` instead of `-1` here?
```
(staticDimSize != ShapedType::kDynamic &&
...
```

File: test/Conversion/TorchToTcp/misc.mlir (hunk 2 above)
Comment (sjain-stanford) [reply test/Conversion/TorchToTcp/misc.mlir:337]: @AaronStGeorge I did it for you here: https://github.com/cruise-automation/mlir-tcp/pull/63 :)

Comment (AaronStGeorge) [reply test/Conversion/TorchToTcp/misc.mlir:337]: Thanks @sjain-stanford :) !

File: lib/Conversion/TorchToTcp/Misc.cpp (hunk 3 above)
Comment (AaronStGeorge) [reply lib/Conversion/TorchToTcp/Misc.cpp:89]: 💯 didn't know about that

Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Misc.cpp:89]: As discussed offline, please ignore since this condition is to check for `-1` in `torch.broadcast_to` or `torch.expand` which stands for "unchanged dim".

File: test/Dialect/misc.mlir [hunk 4]
```diff
@@ -65,6 +65,22 @@ func.func @test_broadcast_axes_w_duplicates(%arg0 : tensor<?x1x?x1xf32>, %arg1 :
...
+  // expected-error@+1{{'tcp.broadcast' op failed to verify that attribute `axes` are in bounds}}
+  %0 = tcp.broadcast %arg0, %arg1, %arg2 {axes = [1, 100]} : tensor<?x1x?x1xf32>, index, index -> tensor<?x?x?x?xf32>
+  return %0 : tensor<?x?x?x?xf32>
+}
+
+// -----
+
+func.func @test_broadcast_axes_not_1(%arg0 : tensor<?x7x?x1xf32>, %arg1 : index, %arg2 : index) -> tensor<?x?x?x?xf32> {
+  // expected-error@+1{{'tcp.broadcast' op failed to verify that dimensions listed in attribute `axes` have a static size of `1`}}
+  %0 = tcp.broadcast %arg0, %arg1, %arg2 {axes = [1, 1]} : tensor<?x7x?x1xf32>, index, index -> tensor<?x?x?x?xf32>
```
Comment (sjain-stanford) [inline test/Dialect/misc.mlir:78]: Perhaps use `axes = [1, 3]` here as `[1, 1]` triggers another category of error (axes must not have any duplicates).

File: test/Conversion/TorchToTcp/misc.mlir [hunk 5]
```diff
@@ -322,3 +322,46 @@ func.func @torch.aten.size.int(%arg0: !torch.vtensor<[?,?],f32>) -> () {
...
+  %int2 = torch.constant.int 2
+  %int4 = torch.constant.int 4
+  %1 = torch.prim.ListConstruct %int4, %int2, %int4, %int2 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
+  %2 = torch.aten.broadcast_to %arg0, %1 : !torch.vtensor<[1,2,1,2],f32>, !torch.list<int> -> !torch.vtensor<[4,2,4,2],f32>
+  return
```
Comment (sjain-stanford) [inline test/Conversion/TorchToTcp/misc.mlir:342]: nit: `return %2 : !torch.vtensor<[4,2,4,2],f32>`

File: lib/Conversion/TorchToTcp/Misc.cpp [hunk 6]
```diff
@@ -81,10 +83,28 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
...
+      bool isDimDynamic =
+          !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize));
+      // lambda because staticDimSize will contain garbage in the case of a
+      // dynamic dim.
+      auto isDimBroadcasted = [i, &inputShape,
+                               newLeadingDims](int64_t dimSize) -> bool {
+        // pytorch defines "passing -1 as the size for a dimension means not
+        // changing the size of that dimension."
+        bool isDimSizePreserved = dimSize == -1;
+        bool doesDimChangeShape = dimSize != inputShape[i - newLeadingDims];
```
Comment (navahgar) [inline lib/Conversion/TorchToTcp/Misc.cpp:103]: This will result in accessing with negative indices when `i < newLeadingDim`, if this lambda is used separately. It is probably not happening now because this lambda is preceded by `isNewDim` condition in your check below.

Since, this is now separated out, I suggest adding the `i >= newLeadingDims` check here as well. Wdyt?

**Review by sjain-stanford — COMMENTED** [review]
Before landing, please rebase / merge main, and enable [this test](https://github.com/cruise-automation/mlir-tcp/blob/main/test/AotCompile/BUILD#L21) in CI.

File: lib/Conversion/TorchToTcp/Misc.cpp (hunk 6 above)
Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Misc.cpp:103]: How about we drop the lambda entirely, and do something like this instead, to handle both the negative indices issue and the garbage staticDimSize issue?
```cpp
bool isNewDim = i < newLeadingDims;
int64_t staticDimSize;
bool isDimDynamic = !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize));
bool isDimSizePreserved = isDimDynamic ? false : staticDimSize == -1;
bool doesDimSizeChange = isNewDim ? false : staticDimSize != inputShape[i - newLeadingDims];
if (isNewDim || isDimDynamic || (!isDimSizePreserved && doesDimSizeChange)) {
...
```

Comment (AaronStGeorge) [reply lib/Conversion/TorchToTcp/Misc.cpp:103]: `doesDimSizeChange` also requires a non dynamic dim. How about just a comment that says the order in which the booleans are checked matters here?

Comment (sjain-stanford) [reply lib/Conversion/TorchToTcp/Misc.cpp:103]: huh I realized that and edited it, but I think GH doesn't send edit emails 😅 
```cpp
bool doesDimSizeChange = (isNewDim || isDimDynamic) ? false : staticDimSize != inputShape[i - newLeadingDims];
```

**Review by navahgar — APPROVED** [review]
LGTM. Thanks for persevering and getting everything fixed properly here.

File: lib/Conversion/TorchToTcp/Misc.cpp [hunk 7]
```diff
@@ -83,10 +85,29 @@ class ConvertAtenBroadcastLikeOps : public OpConversionPattern<AtenOpT> {
...
+      // At the pytorch level it's possible to broadcast to a dynamic dimension;
+      // for example: `torch.broadcast_to(x, y.size())` where `y.size()` has
+      // dynamic shapes. When broadcasting to dynamic size, matchPattern fails
+      // to read an int out of the MLIR Value. In the "usual" case
+      // (`torch.broadcast_to(x, (3, 3))`) matchPattern will succeed.
       int64_t staticDimSize;
-      if (i < newLeadingDims ||
-          !matchPattern(newDimSize, m_TorchConstantInt(&staticDimSize)) ||
-          staticDimSize != -1) {
+      bool isDimDynamic =
```
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Misc.cpp:96]: nit: prefer `isDynamicDim` for consistency with `isNewDim`.

File: lib/Conversion/TorchToTcp/Misc.cpp [hunk 8]
```diff
@@ -83,10 +85,29 @@ class ConvertAtenBroadcastLikeOps : public OpConversionPattern<AtenOpT> {
     SmallVector<Value> resultShape;
     for (int64_t i = 0; i < static_cast<int64_t>(newDimSizes.size()); ++i) {
       Value newDimSize = newDimSizes[i];
+
+      bool isNewDim = i < newLeadingDims;
```
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Misc.cpp:89]: Please add a comment here clarifying PyTorch semantics:
> "Tensor can be also expanded to a larger number of dimensions, and the new ones will be appended at the front. For the new dimensions, the size cannot be set to -1."

File: lib/Dialect/IR/TcpOps.cpp [hunk 9]
```diff
@@ -49,6 +49,22 @@ LogicalResult BroadcastOp::verify() {
...
+    if (axis >= inputShape.size()) {
+      return emitOpError(
+          "failed to verify that attribute `axes` are in bounds");
+    }
+
+    if (inputShape[axis] != 1) {
+      return emitOpError("failed to verify that dimensions listed in attribute "
+                         "`axes` have a static size of `1`");
+    }
+  }
```
Comment (sjain-stanford) [inline lib/Dialect/IR/TcpOps.cpp:66]: Thanks for adding these checks to the verifier!

**Review by sjain-stanford — APPROVED** [review]
Nice catch and thanks for addressing verifier gaps. LGTM besides a few minor naming/documentation requests. 

[non-blocking] In a follow-on , I think we should include tests for the case of `torch.expand` or `torch.broadcast_to` adding new dims (increasing rank).

**Comment by sjain-stanford** [comment]
@AaronStGeorge just saw your vacation ping - please do not bother, I can take over and land this after addressing any remaining nits.
